cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/kafkaassigner/KafkaAssignerDiskUsageDistributionGoal.java

```diff
@@ -0,0 +1,568 @@
+   * @param targetSize the replica size to find.
+   * @param shiftOnExactMatch the position to shift if there is an exact match. The value should be 1 or 0 or -1.
+   *                          When the value is 1, the method returns the position of the first replica whose size is 
+   *                          greater than the target size.
+   *                          When the value is 0, the method returns the position of the first replica whose size is 
+   *                          greater than or equals to the target size.
+   *                          When the value is -1, the method returns the position of the replica whose size is 
+   *                          just less than the target size.
+   *                          
+   * @return the index of the replica whose size is closest to but greater than (or equals to if inclusive=true) the 
+   * target size
+   */
+  private int findReplicaPos(List<ReplicaWrapper> sortedReplicas, double targetSize, int shiftOnExactMatch) {
+    if (shiftOnExactMatch != 1 && shiftOnExactMatch != -1 && shiftOnExactMatch != 0) {
+      throw new IllegalArgumentException("The shift on exact match must be either 1 or -1");
+    }
+    int index = Collections.binarySearch(sortedReplicas, new ReplicaWrapper(null, targetSize),
+                                         Comparator.comparingDouble(ReplicaWrapper::size));
+    switch (shiftOnExactMatch) {
+      case -1:
+        return index >= 0 ? index - 1 : -(index + 1) - 1;
+      case 1:
+        return index >= 0 ? index + 1 : -(index + 1);
+      case 0:
+        return index >= 0 ? index : -(index + 1);
+      default:
+        throw new IllegalStateException("Invalid shift on exact match value " + shiftOnExactMatch);
+    }
+  }
+  
+  private List<Broker> sortedBrokersAboveSizeDescend(ClusterModel clusterModel, double aboveUsage) {
+    List<Broker> sortedBrokersDescend = new ArrayList<>();
+    List<Broker> sortedBrokersAscend = clusterModel.sortedHealthyBrokersUnderThreshold(DISK, Double.MAX_VALUE);
+    for (int i = sortedBrokersAscend.size() - 1; i >= 0; i--) {
+      Broker broker = sortedBrokersAscend.get(i);
+      if (diskUsage(broker) <= aboveUsage) {
+        break;
+      }
+      sortedBrokersDescend.add(sortedBrokersAscend.get(i));
+    }
+    return sortedBrokersDescend;
+  }
+
+  /**
+   * We cannot use {@link Broker#sortedReplicas(Resource)} because the that prioritize the immigrant replicas.
+   */
+  private List<ReplicaWrapper> sortReplicasAscend(Collection<Replica> replicas, Set<String> excludedTopics) {
+    List<ReplicaWrapper> sortedReplicas = new ArrayList<>();
+    replicas.forEach(r -> {
+      if (!excludedTopics.contains(r.topicPartition().topic())) {
+        sortedReplicas.add(new ReplicaWrapper(r, replicaSize(r)));
+      }
+    });
+    sortedReplicas.sort(Comparator.comparingDouble(ReplicaWrapper::size));
+    return sortedReplicas;
+  }
+  
+  private List<Replica> followerReplicas(Broker broker) {
```

Comment:
Would it be more efficient to get leaderReplicas (cached) from the broker and return the set difference from replicas in broker? Is there a reason why we return a list rather than a set?

Reply:
Given that a new list is returned, time complexity wise they are exactly the same. It could be a set, a list is slightly more efficient because the implementation is simpler. In our case, a list is good enough.